### PR TITLE
feat: (Platform) Multi-Combobox component

### PR DESCRIPTION
**Pull request by olshansky**
#### Please provide a link to the associated issue.
Closes: https://github.com/sap/fundamental-ngx/issues/3606
#### Please provide a brief summary of this pull request.
Introduce a new `Multi-Combobox` component.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [n/a] Run npm run build-pack-library and test in external application

Documentation checklist:
- [n/a] update `README.md`
- [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Comment by fkolar** [comment]
What is this WIP about, I dont see any description

**Comment by platon-rov** [comment]
When you start typing the wrong value - the typed character and the dropdown are flickering. In the SAPUI5 if the typed value is wrong - all items are shown, character not flickering.

https://user-images.githubusercontent.com/20265336/114827483-6dc0a700-9dd1-11eb-942b-bc12d122f284.mov

**Comment by puru-hk** [comment]
Attached the accessibility issues 
[Multi Column Issue .docx](https://github.com/SAP/fundamental-ngx/files/6316061/Multi.Column.Issue.docx)

**Comment by platon-rov** [comment]
When moving out from the multi-combobox input if dropdown was opened - it didn't get closed. Should be closed when focus is gone.

**Comment by platon-rov** [comment]
When start typing, gets options, select option with the keyboard - the focus has gone.

When removing (by the backspace) - after removing the last selected item - the focus has gone.

File: apps/docs/src/app/platform/component-docs/platform-multi-combobox/examples/multi-combobox-columns/multi-combobox-columns-example.component.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    selector: 'fdp-multi-combobox-columns-example',
+    templateUrl: './multi-combobox-columns-example.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    providers: [{ provide: DATA_PROVIDERS, useValue: new Map() }]
+})
+export class MultiComboboxColumnsExampleComponent {
+    dataSource = [
+        { name: 'Apple', type: 'Fruits' },
+        { name: 'Banana', type: 'Fruits' },
+        { name: 'Biiiiiiiiiiiiiiiiiiiiiiggggggggggggggggggggggggg Banananananananananananananananananananananananananananananananananananananana', type: 'Fruits' },
```
Comment (platon-rov) [inline apps/docs/src/app/platform/component-docs/platform-multi-combobox/examples/multi-combobox-columns/multi-combobox-columns-example.component.ts:15]: Please use more valuable name :)

File: libs/platform/src/lib/components/form/combobox/combobox.module.ts [hunk 2]
```diff
@@ -20,6 +20,6 @@ import { PlatformAutoCompleteModule } from '../auto-complete/auto-complete.modul
         PlatformAutoCompleteModule
     ],
     providers: [DynamicComponentService],
-    exports: [ComboboxComponent, TemplateModule]
+    exports: [ComboboxComponent, TemplateModule, HighlightPipe]
```
Comment (platon-rov) [inline libs/platform/src/lib/components/form/combobox/combobox.module.ts:23]: I think exposing that pipe to users who will import `comboboxModule` isn't a clear solution. 

By the way we also have `SearchHighlightPipe` which has the same name `| highlight` - could them be combined? If no - won't it conflict where both modules are imported?

File: apps/docs/src/app/platform/component-docs/platform-multi-combobox/platform-multi-combobox-docs.component.html [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    Multi-Combobox with Two Columns
+</fd-docs-section-title>
+<description>
+    Multi-Combobox with two columns of information.
+    Note that a <code>secondaryKey</code> and <code>shoSecondaryText</code> is required
+</description>
+<component-example>
+    <fdp-multi-combobox-columns-example></fdp-multi-combobox-columns-example>
+</component-example>
+<code-example [exampleFiles]="multiComboboxColumnsExample"></code-example>
```
Comment (platon-rov) [inline apps/docs/src/app/platform/component-docs/platform-multi-combobox/platform-multi-combobox-docs.component.html:58]: Groups example here, not columns example.

image

**Comment by platon-rov** [comment]
Text in the second group not properly aligned.

image

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+            this.isOpen = isOpen;
+            this.onTouched();
+            this.openChange.next(isOpen);
+        }
+
+        if (!this.isOpen) {
+            this.searchTermChanged('');
+        }
+
+        this.cd.detectChanges();
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:414]: maybe we need to use `markForCheck` here?

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts [hunk 5]
```diff
@@ -0,0 +1,793 @@
...
+    searchTermChanged(text: string = this.inputText): void {
+        if (text) {
+            this.open();
+        }
+        const map = new Map();
+        map.set('query', text);
+        map.set('limit', 12);
+        this.ds.match(map);
+
+        this.cd.detectChanges();
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:427]: same

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+    }
+
+    /** @hidden */
+    private _getOptionsListWidth(): void {
+        const body = document.body;
+        const rect = (this._element.querySelector('fd-input-group') as HTMLElement).getBoundingClientRect();
+        const scrollBarWidth = body.offsetWidth - body.clientWidth;
+        this.maxWidth = this.autoResize ? window.innerWidth - scrollBarWidth - rect.left : this.minWidth;
+        this.minWidth = rect.width - 2;
+        this._cd.detectChanges();
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:634]: same

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+        }
+    }
+
+    /** @hidden
+     * Method to set input text as item label.
+     */
+    setInputTextFromOptionItem(item: OptionItem): void {
+        if (this.mobile) {
+            this.inputText = item.label;
+            this.cd.detectChanges();
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:229]: same

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+            const selectedItem = this.selectedItems[i];
+            const idx = this._flatSuggestions.findIndex(item => (item.label === selectedItem) || (item.value === selectedItem));
+            if (idx !== -1) {
+                this._selected.push(this._flatSuggestions[idx]);
+                this._flatSuggestions[idx].selected = true;
+            }
+        }
+
+
+        this.cd.detectChanges();
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:316]: same

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+        }
+
+        const activeValue: SelectableOptionItem = this._getSelectItemByInputValue(this.inputText);
+        const index: number = this._flatSuggestions.findIndex(value => value === activeValue);
+        let item: SelectableOptionItem;
+        if (!this.inputText && offset === -1) {
+            item = this._flatSuggestions[this._flatSuggestions.length - 1];
+        } else {
+            item = this._flatSuggestions[index + offset];
+        }
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:516]:
```suggestion
        const position = !this.inputText && offset === -1 ? this._flatSuggestions.length - 1 : index + offset;
        const item: SelectableOptionItem = this._flatSuggestions[position];
```

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts [hunk 10]
```diff
@@ -0,0 +1,793 @@
...
+
+        if (item) {
+            this.setInputTextFromOptionItem(item);
+        }
+
+        const selectedIndex = this._selected.findIndex(value => value.label === item.label);
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:522]: what happens if `item` doesn't exist?

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+    selector: 'fdp-multi-combobox',
+    templateUrl: './multi-combobox.component.html',
+    styleUrls: ['./multi-combobox.component.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    encapsulation: ViewEncapsulation.None,
+    providers: [{ provide: FormFieldControl, useExisting: MultiComboboxComponent, multi: true }]
+})
+export class MultiComboboxComponent extends BaseMultiCombobox implements OnInit, AfterViewInit {
+    @Input()
+    get disabled(): boolean {
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:52]: `@Input disable` already exist in `BaseInput` component
Maybe we need to use it?

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    constructor(
+        readonly cd: ChangeDetectorRef,
+        readonly elementRef: ElementRef,
+        @Optional() @Self() readonly ngControl: NgControl,
+        @Optional() @Self() readonly ngForm: NgForm,
+        @Optional() readonly dialogConfig: DialogConfig,
+        readonly _dynamicComponentService: DynamicComponentService,
+        @Optional() @Inject(DATA_PROVIDERS) private providers: Map<string, DataProvider<any>>,
+        readonly _comboboxConfig: ComboboxConfig,
+        private _rtlService: RtlService,
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:96]: should be Optional

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+        if (!this.dataSource && this.entityClass && providers.has(this.entityClass)) {
+            this.dataSource = new ListDataSource(providers.get(this.entityClass));
+        }
+    }
+
+    /** @hidden */
+    ngAfterViewInit(): void {
+        super.ngAfterViewInit();
+
+        this._rtlService.rtl
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:116]: if Optional then `this._rtlService?.rtl`

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+            .pipe(takeUntil(this._destroyed))
+            .subscribe(isRtl => (this._direction = isRtl ? 'rtl' : 'ltr'));
+
+        if (this._connectedOverlay) {
+            this._connectedOverlay.attach
+                .pipe(takeUntil(this._destroyed))
+                .subscribe(() => this._connectedOverlay.overlayRef.setDirection(this._direction));
+        }
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:124]:
```suggestion
            this._connectedOverlay?.attach
                .pipe(takeUntil(this._destroyed))
                .subscribe(() => this._connectedOverlay.overlayRef.setDirection(this._direction));
```

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+            return;
+        }
+
+        this.inputText = item.label;
+        this.showList(false);
+    }
+
+    /** @hidden */
+    onInputKeyupHandler(): void {
+        if (this.inputText && (!this._suggestions || !this._suggestions.length)) {
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:240]:
```suggestion
        if (this.inputText && !this._suggestions?.length) {
```

**Review by Betrozov — CHANGES_REQUESTED** [review]

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts (hunk 8 above)
Comment (olshansky) [reply libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:316]: @Betrozov here `markForCheck` not checks changes till the change detection will be triggered _(so initially checked items won't be checked until something will be clicked/focused/etc.)_

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts (hunk 10 above)
Comment (olshansky) [reply libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:522]: @Betrozov nothing _(`selectedIndex` will be `-1` and the function do not choose any item)_

Comment (olshansky) [reply libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:522]: @Betrozov with `item?.label` nothing _(`selectedIndex` will be `-1` and the function do not choose any item)_
Thanks, fixed `item?.label`

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Looks good! Just one thing about keyboard support - pressing enter when a search term has autocompleted should select that item and close the popover

**Comment by olshansky** [comment]
> Attached the accessibility issues
> [Multi Column Issue .docx](https://github.com/SAP/fundamental-ngx/files/6316061/Multi.Column.Issue.docx)

@puru-hk part of those issues resolved. The rest is related to components which uses in the Multi-Combobox.
I believe that's not so good idea to modify other components in this PR, but we could raise a separate GitHub issue for it.
Ping me if I needed to raise the issue.

**Comment by InnaAtanasova** [comment]
- The validation part is not working as expected. When you start typing a wrong value, the input is cleared and then after a while the error state disappears.
- when you select an item from the dropdown menu the menu should not close automatically (the user should be able to select multiple items)

**Comment by olshansky** [comment]
> * The validation part is not working as expected. When you start typing a wrong value, the input is cleared and then after a while the error state disappears.

@InnaAtanasova In [SAPUI5 example](https://sapui5.hana.ondemand.com/#/entity/sap.m.MultiComboBox/sample/sap.m.sample.MultiComboBox) it works in a pretty familiar way: if the user typing value which not exists in the list error appears and then after a few seconds disappears. In the [fiori guidelines](https://experience.sap.com/fiori-design-web/multi-combobox/) I haven't found any criteria for how long the invalid state should appear. So, how should it work to be correct?

> * when you select an item from the dropdown menu the menu should not close automatically (the user should be able to select multiple items)

On click to the checkbox or press SPACE, it should not close the dropdown, but on click the label or press ENTER it should close the dropdown. At least this how it works in the SAPUI5 example and partially described in [gitbook](https://app.gitbook.com/@fundamental-ngx/s/docs/approved/wip-multi-combobox#interaction-and-keyboard-support) 🙂

File: libs/platform/src/lib/components/form/multi-combobox/pipes/highlight.pipe.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { MatchingStrategy, MultiComboboxConfig } from '../multi-combobox.config';
+
+// TODO: refactor it with HighlightPipe from Combobox and SearchHighlightPipe
```
Comment (platon-rov) [inline libs/platform/src/lib/components/form/multi-combobox/pipes/highlight.pipe.ts:4]: Is there has issue create ? If not - please create one & put its number here to have a reference.

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox-mobile/multi-combobox/multi-combobox-mobile.component.ts [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+} from '@fundamental-ngx/core';
+
+import { MULTICOMBOBOX_COMPONENT, MultiComboboxInterface } from '../../multi-combobox.interface';
+import { SelectableOptionItem } from '../../../../../domain';
+
+@Component({
+    selector: 'fdp-multi-combobox-mobile',
+    templateUrl: './multi-combobox-mobile.component.html',
+    styles: [`
+        /* TODO: Remove it after fundamental-styles fix release */
```
Comment (platon-rov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox-mobile/multi-combobox/multi-combobox-mobile.component.ts:30]: Has the issue been created? Please provide a reference (#...)

File: libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts [hunk 18]
```diff
@@ -0,0 +1,803 @@
...
+        if (!value) {
+            return;
+        }
+
+        const selectedItems = Array.isArray(value) ? value : [value];
+        super.writeValue(selectedItems);
+    }
+
+    /** @hidden */
+    popoverOpenChangeHandle(): void {
```
Comment (platon-rov) [inline libs/platform/src/lib/components/form/multi-combobox/commons/base-multi-combobox.ts:386]: Seems like the method isn't used anywhere.

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox-mobile/multi-combobox/multi-combobox-mobile.component.ts [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    /** @hidden */
+    ngOnInit(): void {
+        this._listenOnMultiComboboxOpenChange();
+    }
+
+    /** @hidden */
+    ngOnDestroy(): void {
+        super.onDestroy();
```
Comment (platon-rov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox-mobile/multi-combobox/multi-combobox-mobile.component.ts:66]: Do we should have it here? As  the parent lifecycle methods will be called by default if no overrides provided.

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.scss [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+.fdp-multi-combobox__list-container {
+    .fd-list {
+        .fd-list__item {
+            &.fd-list__group-header {
+                padding: 0 1rem;
+            }
+
+            .fd-list__title {
+                // TODO: to investigate
```
Comment (platon-rov) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.scss:19]: Could you please provide more info in comment (in code, not here) about that TODO item? Thanks

**Comment by platon-rov** [comment]
Good job! Left some more small comments but overall it's very good!

**Comment by mikerodonnell89** [comment]
@olshansky build is broken

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox-mobile/multi-combobox/multi-combobox-mobile.component.ts (hunk 19 above)
Comment (olshansky) [reply libs/platform/src/lib/components/form/multi-combobox/multi-combobox-mobile/multi-combobox/multi-combobox-mobile.component.ts:66]: @platon-rov yeah, we should. Unfortunately, `BaseInput.ngOnDestroy` does not execute without `super.onDestroy()` in `MultiCombobox`

**Review by platon-rov — APPROVED** [review]

**Review by Betrozov — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

File: libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+    private _direction: Direction = 'ltr';
+
+    /** @hidden */
+    private _timeout: any; // NodeJS.Timeout
+
+    constructor(
+        readonly cd: ChangeDetectorRef,
+        readonly elementRef: ElementRef,
+        @Optional() @Self() readonly ngControl: NgControl,
+        @Optional() @Self() readonly ngForm: NgForm,
```
Comment (dimamarksman) [inline libs/platform/src/lib/components/form/multi-combobox/multi-combobox/multi-combobox.component.ts:87]: Are you sure we need `@Self` and not `@SkipSelf` in the case of `NgForm`?

**Review by mikerodonnell89 — APPROVED** [review]

**Review by dimamarksman — APPROVED** [review]

**Comment by puru-hk** [comment]
Below example is missing. 
Screenshot 2021-05-20 at 11 29 34

**Comment by puru-hk** [comment]
Tooltip is not shown for dropdown arrow. 

Screenshot 2021-05-20 at 11 47 39

**Comment by puru-hk** [comment]
Looks good

**Review by puru-hk — APPROVED** [review]

**Pull request merged**